scripts/globals/job_utils/ranger.lua

```diff
@@ -0,0 +1,277 @@
+    else
+        action:messageID(target:getID(), xi.msg.basic.JA_MISS_2)
+        action:speceffect(target:getID(), 0)
+    end
+
+    return damage
+end
+
+xi.job_utils.ranger.useVelocityShot = function(player, target, ability)
+	player:addStatusEffect(xi.effect.VELOCITY_SHOT, 1, 0, 7200)
+end
+
+xi.job_utils.ranger.useSharpshot = function(player, target, ability)
+    local power = 40 + player:getMod(xi.mod.SHARPSHOT)
+    player:addStatusEffect(xi.effect.SHARPSHOT, power, 0, 60)
+end
+
+xi.job_utils.ranger.useScavenge = function(player, target, ability)
+    -- RNG AF2 quest check
+    local fireAndBrimstoneCS = player:getCharVar('fireAndBrimstone')
+
+    if
+        player:getZoneID() == xi.zone.CASTLE_OZTROJA and fireAndBrimstoneCS == 5 and-- zone + quest match
+        not player:hasItem(xi.item.OLD_EARRING) and -- make sure player doesn't already have the earring
+        player:getYPos() > -43 and player:getYPos() < -38 and -- Y match
+        player:getXPos() > -85 and player:getXPos() < -73 and -- X match
+        player:getZPos() > -85 and player:getZPos() < -75 and -- Z match
+        math.random(1, 100) <= 50
+    then
+		npcUtil.giveItem(player, xi.item.OLD_EARRING)
+
+    else
+        local bonuses        = (player:getMod(xi.mod.SCAVENGE_EFFECT) + player:getMerit(xi.merit.SCAVENGE_EFFECT)) / 100
+        local arrowsToReturn = math.floor(math.floor(player:getLocalVar('ArrowsUsed') % 10000) * (player:getMainLvl() / 200 + bonuses))
+        local playerID       = target:getID()
+
+        if arrowsToReturn == 0 then
+            action:messageID(playerID, 139)
+        else
+            if arrowsToReturn > 99 then
+                arrowsToReturn = 99
+            end
+
+            local arrowID = math.floor(player:getLocalVar('ArrowsUsed') / 10000)
+            player:addItem(arrowID, arrowsToReturn)
+
+            if arrowsToReturn == 1 then
+                action:messageID(playerID, 140)
+            else
+                action:messageID(playerID, 674)
+                action:additionalEffect(playerID, 1)
+                action:addEffectParam(playerID, arrowsToReturn)
+            end
+
+            player:setLocalVar('ArrowsUsed', 0)
+            return arrowID
+        end
+    end
+end
+
+xi.job_utils.ranger.useCamouflage = function(player, target, ability)
+    local duration = math.random(30, 300) * (1 + 0.01 * player:getMod(xi.mod.CAMOUFLAGE_DURATION))
+    player:addStatusEffect(xi.effect.CAMOUFLAGE, 1 , 0, math.floor(duration * xi.settings.main.SNEAK_INVIS_DURATION_MULTIPLIER))
+end
+
+xi.job_utils.ranger.useBarrage = function(player, target, ability)
+    player:addStatusEffect(xi.effect.BARRAGE, 0, 0, 60)
+end
+
+xi.job_utils.ranger.useShadowbind = function(player, target, ability)
+    if player:getWeaponSkillType(xi.slot.RANGED) == xi.skill.MARKSMANSHIP then -- can't have your crossbow/gun held like a bow, now can we?
+        action:setAnimation(target:getID(), action:getAnimation(target:getID()) + 1)
```

Comment:
Same: Did you test this, action doesn't exist in this scope?

Reply:
I had a test this morning because I was 100% I did test each skill against an enemy. I believe I will need to investigate into the job ability what it should 100% do. The animation triggers and fires the arrow and the target is bound, no arrow was consumed when I had tested this. Will reinvestigate and fix